File: os_py/shared/sys.py
# Copyright (c) 2022, Bamboooz
# All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import platform


def os_name():
    """
    Returns the name of the operating system.
    """
    return platform.system()


def machine_hostname():
    """
    Returns the hostname of the machine.
    """
    return platform.node()


def os_version():
    """
    Returns the version of the operating system.
    """
    return platform.version()


def os_platform():
    """
    Returns the platform of the operating system.
    """
    return platform.platform()


def os_release():
    """
    Returns the release of the operating system.
    """
    return platform.release()


def os_architecture():
    """
    Returns the architecture of the operating system.
    """
    return platform.architecture()[0]

File: ospylib/system/arch/linux/sys.c
/*
Copyright (c) 2022, Bamboooz
All rights reserved.
Use of this source code is governed by a BSD-style license that can be
found in the LICENSE file.
*/

#include <stdio.h>
#include <unistd.h>
#include <string.h>

#define USER_PERMISSION_ADMINISTRATOR 1
#define USER_PERMISSION_NORMAL 0

int admin() {
    return (getuid() == 0);
}

int uptime() {
    FILE * file = fopen("/proc/uptime", "r");
    if (!file) {
        return -1;  // Error occurred
    }

    double uptimeValue;
    if (fscanf(file, "%lf", &uptimeValue) != 1) {
        fclose(file);
        return -1;  // Error occurred
    }

    fclose(file);
    return (int)uptimeValue;
}

int is_vm() {
    FILE * file = fopen("/proc/cpuinfo", "r");
    if (file == NULL) {
        return 0;
    }

    char buffer[256];
    while (fgets(buffer, sizeof(buffer), file) != NULL) {
        if (strstr(buffer, "hypervisor") != NULL) {
            fclose(file);
            return 1;
        }
    }

    fclose(file);
    return 0;
}
